### PR TITLE
Fix IngressClass api version

### PR DESCRIPTION
**Pull request by DirtyCajunRice**
Many automation systems do not template out every sub-api-capability. being v1beta1/v1 should be sufficient to validate usability. To not break functionality and only enhance ability, this PR adds the option of a fallback apiVersion to allow user defined **api version only** of networking.k8s.io

argo-cd templated example (formatted and annotated for readability: 
```
rpc error: code = Unknown desc = Manifest generation error (cached): 
helm template . 
    --name-template traefik-home 
    --namespace ingress 
    --values values.yaml 
    --values values-home.yaml 
    --values /tmp/values-811050304.yaml 
    --api-versions v1 
    --api-versions apiregistration.k8s.io/v1 
    --api-versions apiregistration.k8s.io/v1beta1 
    --api-versions apps/v1 
    --api-versions events.k8s.io/v1 
    --api-versions events.k8s.io/v1beta1 
    --api-versions authentication.k8s.io/v1 
    --api-versions authentication.k8s.io/v1beta1 
    --api-versions authorization.k8s.io/v1 
    --api-versions authorization.k8s.io/v1beta1 
    --api-versions autoscaling/v1 
    --api-versions autoscaling/v2beta1 
    --api-versions autoscaling/v2beta2 
    --api-versions batch/v1 
    --api-versions batch/v1beta1 
    --api-versions certificates.k8s.io/v1 
    --api-versions certificates.k8s.io/v1beta1 
    --api-versions networking.k8s.io/v1         # < has ingressClass
    --api-versions networking.k8s.io/v1beta1    # < has ingressClass
    --api-versions extensions/v1beta1 
    --api-versions policy/v1beta1 
    --api-versions rbac.authorization.k8s.io/v1 
    --api-versions rbac.authorization.k8s.io/v1beta1 
    --api-versions storage.k8s.io/v1 
    --api-versions storage.k8s.io/v1beta1 
    --api-versions admissionregistration.k8s.io/v1 
    --api-versions admissionregistration.k8s.io/v1beta1 
    --api-versions apiextensions.k8s.io/v1 
    --api-versions apiextensions.k8s.io/v1beta1 
    --api-versions scheduling.k8s.io/v1 
    --api-versions scheduling.k8s.io/v1beta1 
    --api-versions coordination.k8s.io/v1 
    --api-versions coordination.k8s.io/v1beta1 
    --api-versions node.k8s.io/v1 
    --api-versions node.k8s.io/v1beta1 
    --api-versions discovery.k8s.io/v1beta1 
    --api-versions flowcontrol.apiserver.k8s.io/v1beta1 
    --api-versions mutators.kubedb.com/v1alpha1 
    --api-versions validators.kubedb.com/v1alpha1 
    --api-versions acme.cert-manager.io/v1 
    --api-versions acme.cert-manager.io/v1beta1 
    --api-versions acme.cert-manager.io/v1alpha3 
    --api-versions acme.cert-manager.io/v1alpha2 
    --api-versions ceph.rook.io/v1 
    --api-versions cert-manager.io/v1 
    --api-versions cert-manager.io/v1beta1 
    --api-versions cert-manager.io/v1alpha3 
    --api-versions cert-manager.io/v1alpha2 
    --api-versions crd.projectcalico.org/v1 
    --api-versions kubernetes-client.io/v1 
    --api-versions monitoring.coreos.com/v1 
    --api-versions monitoring.coreos.com/v1alpha1 
    --api-versions nvidia.com/v1 
    --api-versions appcatalog.appscode.com/v1alpha1 
    --api-versions argoproj.io/v1alpha1 
    --api-versions autoscaling.kubedb.com/v1alpha1 
    --api-versions catalog.kubedb.com/v1alpha1 
    --api-versions externaldns.k8s.io/v1alpha1 
    --api-versions objectbucket.io/v1alpha1 
    --api-versions ops.kubedb.com/v1alpha1 
    --api-versions traefik.containo.us/v1alpha1 
    --api-versions kubedb.com/v1alpha2 
    --api-versions rook.io/v1alpha2 
    --api-versions metrics.k8s.io/v1beta1 
    --include-crds`
failed exit status 1: Error: 
    template: traefik/charts/traefik/templates/ingressclass.yaml:7:8:
        executing "traefik/charts/traefik/templates/ingressclass.yaml" at
            <fail "\n\n ERROR: You must have atleast networking.k8s.io/v1beta1 to use ingressClass">:
                error calling fail: ERROR: You must have atleast networking.k8s.io/v1beta1 to use ingressClass
                    Use --debug flag to render out invalid YAML
```

### CODE REVIEW

File: traefik/templates/ingressclass.yaml [hunk 1]
```diff
@@ -3,6 +3,8 @@
 apiVersion: networking.k8s.io/v1
   {{- else if .Capabilities.APIVersions.Has "networking.k8s.io/v1beta1/IngressClass" }}
 apiVersion: networking.k8s.io/v1beta1
+  {{- else if or (eq .Values.ingressClass.fallbackApiVersion "v1beta1") (eq .Values.ingressClass.fallbackApiVersion "v1") }}
+apiVersion: {{ printf "networking.k8s.io/%s" .Values.ingressClass.fallbackApiVersion }}
```
Comment (Ornias1993) [inline traefik/templates/ingressclass.yaml:7]:
```suggestion
  {{- else if Values.ingressClass.fallbackApiVersion}}
apiVersion: {{ printf "networking.k8s.io/%s" .Values.ingressClass.fallbackApiVersion }}
```

Shouldn't this be fine? Why is it needed to limit this if statement to these two specific fallbackApiVersions?

Comment (DirtyCajunRice) [reply traefik/templates/ingressclass.yaml:7]: @Ornias1993 It is always "fine" to have less validation, but a courtesy check of "did they type v1beta2 on accident" only helps and has no drawbacks.

**Comment by SantoDE** [comment]
Can we do that? Traefik only lists the v1beta version as of now. Does the API Server does some mapping here?

**Comment by DirtyCajunRice** [comment]
Yes the api does a conversion. And as much as id love for traefik to catch up on the ingress + ingressclass listeners this can happen much more quickly. I currently use v1 manually deployed and have the ingressclass from the chart disabled. cluster is 1.20.5 and works perfectly.

**Comment by neuromantik33** [comment]
I have exactly the same problem running with helm 3.4.5 and argo 2.0.1. Obviously manually creating the ingress class is a solution, but I'm looking for something less patch-like.

**Comment by SantoDE** [comment]
We will merge that for sure :)

I've also just raised https://github.com/traefik/traefik/pull/8089 to correct the behaviour in Traefik.

**Comment by SantoDE** [comment]
@DirtyCajunRice just to be sure: by default, we stay with `networking.k8s.io/v1beta1` but we can overwrite that, right?

**Comment by Ornias1993** [comment]
> @DirtyCajunRice just to be sure: by default, we stay with `networking.k8s.io/v1beta1` but we can overwrite that, right?

You could've taken the effort to click on the code tab and read the four-liner ;-)

```
apiVersion: networking.k8s.io/v1
  {{- else if .Capabilities.APIVersions.Has "networking.k8s.io/v1beta1/IngressClass" }}
apiVersion: networking.k8s.io/v1beta1
  {{- else if or (eq .Values.ingressClass.fallbackApiVersion "v1beta1") (eq .Values.ingressClass.fallbackApiVersion "v1") }}
apiVersion: {{ printf "networking.k8s.io/%s" .Values.ingressClass.fallbackApiVersion }}
```

aka: Currently the default isn't even `networking.k8s.io/v1beta1` ;-)

**Comment by SantoDE** [comment]
I tried to understand the line, thats why I asked :)

Thanks for clarifying! :)

**Review by SantoDE — APPROVED** [review]
LGTM

**Pull request merged**